src/server/pfs/server/driver_fsck.go

```diff
@@ -134,163 +134,224 @@ func (e ErrMissingBranchHead) Error() string {
 	return fmt.Sprintf("consistency error: branch %s does not have a head commit", e.Branch)
 }
 
-// fsck verifies that pfs satisfies the following invariants:
-// 1. Branch provenance is transitive
-// 2. Head commit provenance has heads of branch's branch provenance
-// If fix is true it will attempt to fix as many of these issues as it can.
-func (d *driver) fsck(ctx context.Context, fix bool, cb func(*pfs.FsckResponse) error) error {
-	onError := func(err error) error { return cb(&pfs.FsckResponse{Error: err.Error()}) }
-
-	// TODO(global ids): no fixable fsck issues?
-	// onFix := func(fix string) error { return cb(&pfs.FsckResponse{Fix: fix}) }
+func checkBranchProvenances(bi *pfs.BranchInfo, branchInfos map[string]*pfs.BranchInfo, onError func(error) error) error {
+	// we expect the branch's provenance to equal the union of the provenances of the branch's direct provenances
+	// i.e. union(branch, branch.Provenance) = union(branch, branch.DirectProvenance, branch.DirectProvenance.Provenance)
+	direct := bi.DirectProvenance
+	union := []*pfs.Branch{bi.Branch}
+
+	for _, directProvenance := range direct {
+		directProvenanceInfo := branchInfos[pfsdb.BranchKey(directProvenance)]
+		union = append(union, directProvenance)
```

Comment:
We probably shouldn't append when directProvenance is nil.

Reply:
do you mean when `directProvenanceInfo` is `nil`?

Reply:
done